Add Viewer for large coverart on click

defvs opened the pull request.

Xerus2000 commented on src/main/xerus/monstercat/api/Player.kt:
```diff
@@ -169,7 +176,25 @@ object Player: FadingHBox(true, targetHeight = 25) {
+				val imageView = ImageView(Covers.getCoverImage(coverUrl!!, 24))
+				imageView.setOnMouseClicked {
+					if (it.button == MouseButton.PRIMARY && it.clickCount == 1){
+						val size: Double = minOf(Screen.getPrimary().visualBounds.height, Screen.getPrimary().visualBounds.height) / 2
```
min of two equal values? Typo?

defvs replied:
Yes, typo. Should be .width and .height

Xerus2000 commented on src/main/xerus/monstercat/api/Covers.kt:
```diff
@@ -32,9 +32,16 @@ object Covers {
+			getLargeCover(coverUrl, size).use { Image(it, size.toDouble(), size.toDouble(), false, false) }
+	
+	fun getLargeCover(coverUrl: String, size: Int): InputStream =
+			fetchCover(coverUrl, size).content
```
Since we now use the large covers in multiple places, we should consider caching them as well, or?

defvs replied:
If we use the large covert anywhere else, why not, but this will take massive disk space after a while.

Xerus2000 replied:
That's why we need to automatically prune the cache. There's already such a mechanism for the logs.

Xerus2000 replied:
but that's not urgent

defvs replied:
Well it's obsolete; I cache the releases now

Xerus2000 replied:
huh?

defvs replied:
The large covers are cached like the thumbnails now. Marking as resolved

defvs commented:
Fixed the typo;
Added caching for large file.
Could do some improvements on the fuctions, but at this point other than codestyle it's ready to merge.

defvs commented:
The already existing cover images has been renamed "thumbnail" as it is a very small 64x64 image, thus being a thumbnail. The "large" cover art is now the real cover.

defvs commented:
Closing; Already merged in #15 because needed.

Xerus2000 commented:
Why needed? I don't like it if PRs grow unnecessarily big, and these seem to be two distinct features.

defvs commented:
I needed a fix for my Playlist branch. You could still merge this before #15 if you want it now.

Xerus2000 commented on src/main/xerus/monstercat/downloader/SongView.kt:
```diff
@@ -163,9 +163,9 @@ class SongView(private val sorter: ObservableValue<ReleaseSorting>):
+				var image = Covers.getCoverThumbnail(release.coverUrl, 16)
 				if(image.exception != null) {
-					image = Covers.getCoverImage(release.coverUrl, 16, true)
+					image = Covers.getCoverThumbnail(release.coverUrl, 16, true)
```
pls just call it getThumbnail

Xerus2000 replied:
or maybe getThumbnailImage

defvs replied:
I'm more on getThumbnail

defvs replied:
I'll refactor theses function a bit

defvs replied:
Fixed

Xerus2000 commented on src/main/xerus/monstercat/api/response/Release.kt:
```diff
@@ -41,6 +43,17 @@ data class Release(
+	 * Can be used after a track fetch.
+	 * [tracks] Tracks corresponding to that release which needs to be linked.
+	 */
+	fun updateTracks(tracks: List<Track>) {
```
what is this for?

defvs replied:
Tracks fetched from the cache are not linked to Releases in the init. This is a workaround to force-link the tracks and the releases.

Xerus2000 commented on src/main/xerus/monstercat/api/response/Track.kt:
```diff
@@ -53,15 +50,15 @@ open class Track: MusicItem() {
 			return this
 		
-		if(::release.isInitialized) {
+		/*if(::release.isInitialized) { TODO : Remove this as it's useless, unless proven
```
are you sure?

defvs replied:
Linked to nothing.

Xerus2000 replied:
Linked? You mean it was never called?

defvs replied:
Yeah, only in TabDownloader, where it is set but never read.

defvs replied:
You can add it back, but you'll have to deal with a crash I couldn't  find the cause

Xerus2000 replied:
Have you sent me the stacktrace? I never got a crash.

Xerus2000 commented:
I'd like to squash this PR since it's rather small with many trivial commits, but that would create conflicts with the other branch... That's why I don't like intertwining branches.

defvs commented:
Which branch ? If it is Playlist, I'll just revert back before my merge.

defvs commented on src/main/xerus/monstercat/MonsterUtilities.kt:
```diff
@@ -307,4 +311,71 @@ class MonsterUtilities(checkForUpdate: Boolean): VBox(), JFXMessageDisplay {
+			height = windowSize
+			width = windowSize
+		}
+		stage.widthProperty().addListener { observable, oldValue, newValue ->
```
Should be removed.

defvs replied:
4f4d1de7bbed7a0797b3ac2ed6282aacb83f4198

defvs commented on src/main/xerus/monstercat/MonsterUtilities.kt:
```diff
@@ -307,4 +311,71 @@ class MonsterUtilities(checkForUpdate: Boolean): VBox(), JFXMessageDisplay {
+		stage.apply {
+			this.isResizable = isResizable
+			
+			widthProperty().addListener { observable, oldValue, newValue ->
```
Rename observable, oldValue to _

defvs replied:
4f4d1de7bbed7a0797b3ac2ed6282aacb83f4198

Xerus2000 commented on src/main/xerus/monstercat/api/Covers.kt:
```diff
@@ -5,24 +5,29 @@ import org.apache.http.HttpEntity
 object Covers {
 	
 	private val coverCacheDir = cacheDir.resolve("cover-images").apply { mkdirs() }
+	private val largeCoverCacheDir = cacheDir.resolve("large-cover-images").apply { mkdirs() }
```
please rename the dir variables to match the function names, i.e. thumbnail and cover

defvs replied:
Renamed in 5a080655c024620044533366bd60b61dc00493df

Xerus2000 commented on src/main/xerus/monstercat/api/response/Track.kt:
```diff
@@ -53,16 +50,6 @@ open class Track: MusicItem() {
 		if(titleClean.isNotEmpty())
 			return this
 		
-		if(::release.isInitialized) {
```
Even if we were to remove this, it does not belong into this PR. This is supposed to work, so I'm gonna fix it.

defvs replied:
Merged your implementation in b6a24c5a38d14eba3594c76b12502c2d720a1724

Xerus2000 commented on src/main/xerus/monstercat/MonsterUtilities.kt:
```diff
@@ -310,4 +314,64 @@ class MonsterUtilities(checkForUpdate: Boolean): VBox(), JFXMessageDisplay {
 	}
 	
+	/** Shows a new window with an ImageView of the requested [coverUrl]
+	 * [title] Title of the window, only useful when decorated
```
please use `@param` instead of links

defvs replied:
ce50572db5b3eda667478e18f6d1c9fd1d0584b0

Xerus2000 commented on src/main/xerus/monstercat/api/Covers.kt:
```diff
@@ -33,8 +38,23 @@ object Covers {
-	private fun coverCacheFile(coverUrl: String) =
-		coverCacheDir.apply { mkdirs() }.resolve(coverUrl.substringAfterLast('/').replaceIllegalFileChars())
+	/** Returns a larger Image of the cover in the requested size using caching.
+	 * @param size the size of the Image - the downloaded image will be at maximum 2048x2048 */
```
```suggestion
	 * @param size the size of the Image - the image file will always be 2048x2048 */
```

defvs replied:
677fca2e9aaa170bb1cb90c3ce7db6e81d93cae8

Xerus2000 commented on src/main/xerus/monstercat/api/Covers.kt:
```diff
@@ -33,8 +38,23 @@ object Covers {
+			getCover(coverUrl, invalidate).use { createImage(it, size) }
+	
+	/** Returns an InputStream to the cover in size 2048x2048, using caching. */
+	private fun getCover(coverUrl: String, invalidate: Boolean = false) : InputStream {
```
Please merge this method with the `getThumbnail` method, they are almost equal. DRY!

defvs replied:
I cannot do that, because both use different caching directories. I'd need to add a new argument "thumbnail: Boolean" and it will end up being one big function with an if(thumbnail) getThumbnail else getCover

Xerus2000 replied:
1. They don't have to use different directories. We could also adjust the filename pattern to include the size.
2. The only other thing that is different is the size of the cover. That could be a parameter.

Yes I can also do this after the merge, but I'd like you to learn to recognise these things by yourself ;)

defvs replied:
imo we should keep them in separate directories. Makes for easier cleanup by hand when folders start to get huge. I'll add a check, if image size is above 128x128 it's not a thumbnail anymore.

defvs replied:
fcb4c904537d730101db6cc454c453171c86d351

Xerus2000 commented on src/main/xerus/monstercat/api/Player.kt:
```diff
@@ -169,7 +171,13 @@ object Player: FadingHBox(true, targetHeight = 25) {
-				children.add(0, ImageView(Covers.getCoverImage(coverUrl!!, 24)))
+				val imageView = ImageView(Covers.getThumbnailImage(coverUrl!!, 24))
+				imageView.setOnMouseClicked {
+					if (it.button == MouseButton.PRIMARY && (it.clickCount == 1 || it.clickCount == 2)) {
```
why is the clickCount relevant?

defvs replied:
e8858dcd7c9c1e6f0658fe55c1781fcfe0417165

Xerus2000 commented on src/main/xerus/monstercat/MonsterUtilities.kt:
```diff
@@ -310,4 +314,65 @@ class MonsterUtilities(checkForUpdate: Boolean): VBox(), JFXMessageDisplay {
+					largeImage.fitWidth = newValue
+				}
+			}
+			this.isResizable = isResizable
```
what is this?

Xerus2000 commented on src/main/xerus/monstercat/MonsterUtilities.kt:
```diff
@@ -310,4 +314,65 @@ class MonsterUtilities(checkForUpdate: Boolean): VBox(), JFXMessageDisplay {
+			}
+			this.isResizable = isResizable
+			
+			widthProperty().addListener { _, _, newValue ->
```
is there a reason you add this listener twice?

defvs replied:
Removed in 46b6182eb6f7a6859969136b3bcbfd548962ddbb

Xerus2000 commented on src/main/xerus/monstercat/MonsterUtilities.kt:
```diff
@@ -310,4 +314,65 @@ class MonsterUtilities(checkForUpdate: Boolean): VBox(), JFXMessageDisplay {
+			}
+			
+			if (closeOnFocusLost) {
+				focusedProperty().addListener { _, lostFocus, _ ->
```
This doesn't look correct. You are calling the old value of focused "lostFocus", but it should be "oldFocus". This might currently work, but is semantically completely incorrect.

defvs replied:
It's true; if oldFocus is 'true', that means that we are no longer focused; Just replace it with !newFocus

defvs replied:
Fixed in 1f25196964e7dc2b6a3ef6372af5880f3ec0b6ff

Xerus2000 replied:
Practically it worked like this, but theoretically there could be a case where both oldFocus and newFocus are true, which would make this method falsely trigger. Anyways, it's good now :)

defvs commented:
Travis keeps failing, no idea why. semaphore works fine, local build too.

Xerus2000 commented on src/main/xerus/monstercat/tabs/TabSettings.kt:
```diff
@@ -80,6 +83,16 @@ class TabSettings: VTab() {
 			Settings.PLAYERSEEKBARHEIGHT.bind(valueProperty() as ObservableValue<out Double>)
 		})
 		
+		val coverPriority = ComboBox(FXCollections.observableArrayList<PriorityList>(*PriorityList.values())).apply {
```
Please have a look at the `createComboBox` method, it seems appropriate for this use-case.

defvs replied:
ab6d22c145cef162f1c37730eb8e620acf17e1bd

Xerus2000 commented on src/main/xerus/monstercat/tabs/TabSettings.kt:
```diff
@@ -224,4 +237,33 @@ class TabSettings: VTab() {
+			}
+			
+			fun findFromList(list: List<String>): PriorityList{
+				PriorityList.values().forEach {
```
For these methods, you can use the `find` function instead of forEach, and then add the default value via `?:`. That way you could even write the function in expression syntax.

defvs replied:
5773bcb0210f66cd9a75b2182376218ba739f728

Xerus2000 commented:
> Travis keeps failing, no idea why. semaphore works fine, local build too.
> Seems like the build cannot find "Release.Type.MIXES". Could be a conflict ? It finds all the other types fine.

It is now called "MIX", I renamed it for consistency since everything else is in singular as well. I am actually confused why it works locally for you and on Semaphore ^^

defvs commented:
Because it is still named "MIXES" on this branch ? We're not up to date on master.

Xerus2000 commented:
> Because it is still named "MIXES" on this branch ? We're not up to date on master. Probably because Travis merges this branch into master THEN compiles !

Ah right, yes, this is the PR build, which means it checks whether it works when merged.

defvs commented:
I will merge master into it to resolve conflicts after we're done.

defvs commented:
Merged master; ready to merge.

Xerus2000 commented on src/main/xerus/monstercat/api/Covers.kt:
```diff
@@ -5,26 +5,35 @@ import org.apache.http.HttpEntity
+	 * @param invalidate set to true to ignore already existing cache files
+	 */
+	fun getCover(coverUrl: String, size: Int = 64, invalidate: Boolean = false) : InputStream {
+		val file = coverCacheFile(if (size > 128) coverCacheDir else thumbnailCacheDir, coverUrl)
```
I'd just use the same directory and then pass the size to this method so it is appended to the filename, i.e. "coverx64.jpg"

defvs replied:
Have you read my answer above ?

Xerus2000 replied:
oh, it got lost. Then we'll leave it like that for now, maybe I reconsider this in the future.

defvs replied:
I actually implemented it. Up to you to revert it or not. adc75cd97b5ecc7555fab0df8d66264266e00984

Xerus2000 commented:
Please remove the commented-out code-blocks and fix this little comment, other than that LGTM.

defvs commented:
Yeah totally forgot about my commented block.

defvs commented:
Got no time till tonight. I'm still trying to figure out a way to change the filename to end with size while the extension is left untouched

Xerus2000 commented:
Kotlin has an extensions function to get the `extension` and the files's `nameWithoutExtension` ;)

defvs commented:
DO NOT MERGE ! There is an issue where the program will freeze when playing a song (even when resolving conflicts)

defvs commented:
Fixed it in cf9cc81b3ac8c11cc540821d41e1dada92c52c9b if you want to cherrypick into master

defvs commented:
Ready to merge.

Xerus2000 commented on src/main/xerus/monstercat/api/Player.kt:
```diff
@@ -231,9 +239,12 @@ object Player: FadingHBox(true, targetHeight = 25) {
-		checkFx {
-			backgroundCover.value = coverUrl?.let {
-				Background(BackgroundImage(Covers.getCoverImage(coverUrl), BackgroundRepeat.NO_REPEAT, BackgroundRepeat.NO_REPEAT, BackgroundPosition.CENTER, BackgroundSize(100.0, 100.0, true, true, true, true)))
+		GlobalScope.launch {
```
I think we can avoid this change if we simply enable backgroundLoading for the Image created in `createImage`. SHould be a constructor parameter.

Xerus2000 replied:
nevermind, doesn't work for InputStreams

Xerus2000 commented:
I have solved the problems in "Covers" and incorporated your changes into master. It should be possible to resolve the conflicts by simply accepting my changes. please have a look at my comments nonetheless, and how I resolved them in my commit.

defvs commented on src/main/xerus/monstercat/api/Covers.kt:
```diff
@@ -5,26 +5,37 @@ import org.apache.http.HttpEntity
-	fun getCoverImage(coverUrl: String, size: Number = 64, invalidate: Boolean = false): Image =
-		getCover(coverUrl, invalidate).use { createImage(it, size) }
+	private fun coverCacheFile(file: File, coverUrl: String, size: Int) = file.run {
+			mkdir()
```
yeah it probably is a typo, as I remember looking at the differences. Will fix thanks.

defvs replied:
a8028198af4a6e176c9750110508466891e593e4 as well

defvs replied:
hotfix 0bbebfaebe04649ffdc499a6227cec2dd0cb44bd

defvs commented on src/main/xerus/monstercat/api/Covers.kt:
```diff
@@ -5,26 +5,37 @@ import org.apache.http.HttpEntity
 	
 	private fun createImage(content: InputStream, size: Number) =
-		Image(content, size.toDouble(), size.toDouble(), false, false)
+			Image(content, size.toDouble(), size.toDouble(), false, false)
```
Garbage IDE :laughing:

defvs replied:
Fixed in 9ebc6eccf9167a194dd295e1019631ba714ca4eb

defvs commented on src/main/xerus/monstercat/api/Covers.kt:
```diff
@@ -33,8 +44,10 @@ object Covers {
+	/** Returns a larger Image of the cover in the requested size using caching.
+	 * @param size the size of the Image - the image file will always be 2048x2048 */
+	fun getCoverImage(coverUrl: String, size: Int = 2048, invalidate: Boolean = false): Image =
+			getCover(coverUrl, 2048, invalidate).use { createImage(it, size) }
```
Sure.

defvs replied:
3406011d032c0486214104673065ad027c9ca922

defvs commented:
Good, thanks.

defvs commented on src/main/xerus/monstercat/api/Covers.kt:
```diff
@@ -5,26 +5,37 @@ import org.apache.http.HttpEntity
-	 * @param size the size of the Image - the underlying image data will always be 64x64, thus this is the default. */
-	fun getCoverImage(coverUrl: String, size: Number = 64, invalidate: Boolean = false): Image =
-		getCover(coverUrl, invalidate).use { createImage(it, size) }
+	private fun coverCacheFile(file: File, coverUrl: String, size: Int) = file.run {
```
a8028198af4a6e176c9750110508466891e593e4
But is it better to use let in theses cases ?

defvs replied:
0bbebfaebe04649ffdc499a6227cec2dd0cb44bd for hotfix

defvs commented on src/main/xerus/monstercat/api/Covers.kt:
```diff
@@ -5,26 +5,37 @@ import org.apache.http.HttpEntity
+
+	/** @return an [InputStream] of the downloaded image file
+	 * @param coverUrl the URL for fetching the cover
+	 * @param size the final size of the cover
```
9a68ae1ff6a64314a5f61ad51c4980cad9363f94 is this better ?

defvs commented:
Wait, just understood, all the comments here, you fixed them on master ?

defvs commented:
Merged master just fine. Thanks

Xerus2000 commented on src/main/xerus/monstercat/api/Player.kt:
```diff
@@ -200,6 +207,7 @@ object Player: FadingHBox(true, targetHeight = 25) {
 		updateCover(null)
 		disposePlayer()
 		GlobalScope.launch {
+			updateCover(null)
```
no?

defvs replied:
Merge conflicts which were badly resolved.

defvs replied:
61113549ea3fcc29a9d55f4590548a42a20a0919

Xerus2000 commented on src/main/xerus/monstercat/api/Covers.kt:
```diff
@@ -55,18 +55,13 @@ object Covers {
 	/** Attaches a parameter to the [coverUrl] for the requested [size] */
-	private fun getCoverUrl(coverUrl: String, size: Int? = null) =
-		URI(coverUrl).toASCIIString() + size?.let { "?image_width=$it" }.orEmpty()
+	private fun getCoverUrl(coverUrl: String, size: Int? = null) = URI(coverUrl).toASCIIString() + size?.let { "?image_width=$it" }.orEmpty()
```
Please revert this reformat

defvs replied:
b66ddfdb8e02530d690654b7171fad61ff5e6bb2

Xerus2000 commented on src/main/xerus/monstercat/api/Cache.kt:
```diff
@@ -38,10 +38,15 @@ object Cache: Refresher() {
 	}
 	
-	/** Gets all tracks by flatMapping all the tracks of all Releases */
+	/** Gets all tracks by flatMapping all the tracks of all Releases. Two tracks with the same id will only yield one (hashSet) */
```
```suggestion
	/** Gets a complete collection of unique Tracks. */
```

defvs commented on src/main/xerus/monstercat/api/Cache.kt:
```diff
@@ -38,10 +38,15 @@ object Cache: Refresher() {
 		getReleases().flatMap { it.tracks }.toHashSet()
 	
+	/** Gets all tracks by flatMapping all the tracks of all Releases. Will return all tracks, regardless of id duplicates.
+	 * This is very useful when you need a separate Track for each Release it is part of.*/
```
This is vewy vewy useful :smile_cat:
